File: packages/akilent-python/src/akilent/resources.py
"""Resource namespaces exposed on the client (`client.messages`, …)."""
from __future__ import annotations

from typing import Any, Iterator, List, Optional

from ._transport import Transport


class _Base:
    def __init__(self, transport: Transport) -> None:
        self._t = transport


class Messages(_Base):
    def send(
        self,
        *,
        from_: str,
        to: str,
        subject: str = "",
        text: Optional[str] = None,
        html: Optional[str] = None,
        template: Optional[str] = None,
        variables: Optional[dict] = None,
        locale: Optional[str] = None,
        attachments: Optional[list] = None,
        idempotency_key: Optional[str] = None,
    ) -> dict:
        payload: dict[str, Any] = {"from": from_, "to": to}
        if subject:
            payload["subject"] = subject
        if text is not None:
            payload["text"] = text
        if html is not None:
            payload["html"] = html
        if template is not None:
            payload["template"] = template
        if variables is not None:
            payload["template_variables"] = variables
        if locale is not None:
            payload["locale"] = locale
        if attachments is not None:
            payload["attachments"] = attachments
        return self._t.request(
            "POST", "/api/v1/messages", json=payload, idempotency_key=idempotency_key
        )

    def list(self, **filters: Any) -> dict:
        return self._t.request("GET", "/api/v1/messages", params=filters or None)

    def retrieve(self, message_id: str) -> dict:
        return self._t.request("GET", f"/api/v1/messages/{message_id}")

    def events(self, message_id: str) -> list:
        return self._t.request("GET", f"/api/v1/messages/{message_id}/events")["data"]

    def iter(self, *, page_size: int = 100, **filters: Any) -> Iterator[dict]:
        offset = 0
        while True:
            page = self.list(limit=page_size, offset=offset, **filters)
            rows: List[dict] = page.get("data", [])
            for row in rows:
                yield row
            offset += len(rows)
            if not rows or offset >= page.get("total", 0):
                return


class Templates(_Base):
    def list(self) -> list:
        return self._t.request("GET", "/api/v1/templates")

    def create(self, **body: Any) -> dict:
        return self._t.request("POST", "/api/v1/templates", json=body)

    def retrieve(self, slug: str) -> dict:
        return self._t.request("GET", f"/api/v1/templates/{slug}")

    def update(self, slug: str, **body: Any) -> dict:
        return self._t.request("PATCH", f"/api/v1/templates/{slug}", json=body)

    def delete(self, slug: str) -> None:
        self._t.request("DELETE", f"/api/v1/templates/{slug}")

    def preview(self, slug: str, *, variables: Optional[dict] = None) -> dict:
        return self._t.request(
            "POST", f"/api/v1/templates/{slug}/preview", json={"variables": variables or {}}
        )

    def clone(self, slug: str) -> dict:
        return self._t.request("POST", f"/api/v1/templates/{slug}/clone")


class Campaigns(_Base):
    def create(self, **body: Any) -> dict:
        return self._t.request("POST", "/api/v1/campaigns", json=body)

    def retrieve(self, campaign_id: int) -> dict:
        return self._t.request("GET", f"/api/v1/campaigns/{campaign_id}")


class RequestLogs(_Base):
    def list(self, **filters: Any) -> dict:
        return self._t.request("GET", "/api/v1/request-logs", params=filters or None)

    def retrieve(self, request_id: str) -> dict:
        return self._t.request("GET", f"/api/v1/request-logs/{request_id}")

File: packages/akilent-node/src/resources.ts
import type { Transport } from "./client.js";

export interface SendParams {
  from: string;
  to: string;
  subject?: string;
  text?: string;
  html?: string;
  template?: string;
  variables?: Record<string, unknown>;
  locale?: string;
  attachments?: Array<{ filename: string; content_b64: string; content_type?: string }>;
  idempotencyKey?: string;
}

export interface Message {
  id: string;
  status: string;
  from: string;
  to: string;
  subject: string;
  template: string | null;
  campaign_id: number | null;
  created_at: string;
  sent_at: string | null;
}

export interface MessageEvent {
  id: string;
  type: string;
  source: string;
  occurred_at: string;
  data: Record<string, unknown>;
  request_id: string | null;
}

interface Page<T> {
  data: T[];
  total: number;
  limit: number;
  offset: number;
}

export class Messages {
  constructor(private readonly t: Transport) {}

  send(p: SendParams): Promise<{ id: number; public_id: string; status: string }> {
    const { idempotencyKey, variables, ...rest } = p;
    const body: Record<string, unknown> = { ...rest };
    if (variables !== undefined) body.template_variables = variables;
    return this.t.request("POST", "/api/v1/messages", { body, idempotencyKey });
  }

  list(filters: Record<string, string | number> = {}): Promise<Page<Message>> {
    return this.t.request("GET", "/api/v1/messages", { params: filters });
  }

  retrieve(id: string): Promise<Message & { events: MessageEvent[] }> {
    return this.t.request("GET", `/api/v1/messages/${id}`);
  }

  async events(id: string): Promise<MessageEvent[]> {
    const r = await this.t.request<{ data: MessageEvent[] }>(
      "GET",
      `/api/v1/messages/${id}/events`,
    );
    return r.data;
  }

  async *iterate(
    filters: Record<string, string | number> = {},
    pageSize = 100,
  ): AsyncGenerator<Message> {
    let offset = 0;
    for (;;) {
      const page = await this.list({ ...filters, limit: pageSize, offset });
      for (const m of page.data) yield m;
      offset += page.data.length;
      if (page.data.length === 0 || offset >= page.total) return;
    }
  }
}

export class Templates {
  constructor(private readonly t: Transport) {}
  list(): Promise<unknown[]> {
    return this.t.request("GET", "/api/v1/templates");
  }
  create(body: Record<string, unknown>): Promise<unknown> {
    return this.t.request("POST", "/api/v1/templates", { body });
  }
  retrieve(slug: string): Promise<unknown> {
    return this.t.request("GET", `/api/v1/templates/${slug}`);
  }
  update(slug: string, body: Record<string, unknown>): Promise<unknown> {
    return this.t.request("PATCH", `/api/v1/templates/${slug}`, { body });
  }
  remove(slug: string): Promise<void> {
    return this.t.request("DELETE", `/api/v1/templates/${slug}`);
  }
  preview(slug: string, variables: Record<string, unknown> = {}): Promise<{
    subject: string;
    html: string;
    text: string;
    missing_variables: string[];
  }> {
    return this.t.request("POST", `/api/v1/templates/${slug}/preview`, {
      body: { variables },
    });
  }
  clone(slug: string): Promise<unknown> {
    return this.t.request("POST", `/api/v1/templates/${slug}/clone`, { body: {} });
  }
}

export class Campaigns {
  constructor(private readonly t: Transport) {}
  create(body: Record<string, unknown>): Promise<unknown> {
    return this.t.request("POST", "/api/v1/campaigns", { body });
  }
  retrieve(id: number): Promise<unknown> {
    return this.t.request("GET", `/api/v1/campaigns/${id}`);
  }
}

export class RequestLogs {
  constructor(private readonly t: Transport) {}
  list(filters: Record<string, string | number> = {}): Promise<unknown> {
    return this.t.request("GET", "/api/v1/request-logs", { params: filters });
  }
  retrieve(requestId: string): Promise<unknown> {
    return this.t.request("GET", `/api/v1/request-logs/${requestId}`);
  }
}
